### ISSUE TITLE
[Bug]: Get-DbaDbBackupHistory  (Break in LSN chain)

### ISSUE THREAD

**Issue by saxet77** [dataplat/dbatools#7515]
<!--
🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨
The core team may close bug reports that do not follow our provided template.

The information we ask for helps to better troubleshoot the report. We release frequently and often, this information helps to resolve the issue more efficiently.
-->

### Report

<!--
Things to consider/verify:
- Errors received are not related to permissions? NO
- Have you tried the command using `powershell.exe`? YES
- If `Copy-DbaDatabase`, can you replicate the problem with `Backup-DbaDatabase` and `Restore-DbaDatabase`?
- `Copy-DbaDatabase` will not work in every environment and every situation.
-->

#### Host used

- [ ] powershell.exe
- [x] ISE
- [ ] VS Code
- [ ] Other (please specify)

#### Errors Received
<!--
🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨
Run the command and paste the output of `$error[0] | select *` below
-->
```

WARNING: [21:10:15][Restore-DbaDatabase] Database TODELETE failed testing,  skipping
```
#### Steps to Reproduce
```
# Get backup history from the SourceInstance msdb database
$ListBackupFile = Get-DbaDbBackupHistory -SqlInstance Server1 -Database Database1 -last 
  
 # Generate the PIT restore script
 $RestoreScript = $ListBackupFile | Restore-DbaDatabase -SqlInstance Server2 -DatabaseName TODELETE -RestoreTime '2021-07-07 01:17:15' -WithReplace -OutputScriptOnly -ReplaceDbNameInFile -TrustDbBackupHistory;

$ListBackupFile
SqlInstance                 Database   Type TotalSize   DeviceType Start                   Duration End                    
-----------                 --------   ---- ---------   ---------- -----                   -------- ---                    

#### Expected Behavior
```
Restore Script provide by SSMS 'Restore/Database/Script'

#### Actual Behavior

Restore Script provide by DBATools (missing 1 log file)


### Environmental information

🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨
Please provide the output of the below script

```powershell
Powershell Version = 5.1.14393.3866
dbatools lastest installed = 1.0.169
Culture of OS = 
LCID             Name             DisplayName                                                                                                                                                                                                                              
----             ----             -----------                                                                                                                                                                                                                              
1036             fr-FR            French (France)
```

#### SQL Server:

```sql
Microsoft SQL Server 2017 (RTM-CU23) (KB5000685) - 14.0.3381.3 (X64)   Feb  9 2021 12:08:50   Copyright (C) 2017 Microsoft Corporation  Developer Edition (64-bit) on Windows Server 2016 Datacenter 10.0 <X64> (Build 14393: ) (Hypervisor) 
```

```sql
us_english
```

**Comment by saxet77** [comment]
Expected Behavior
RESTORE DATABASE [TODELETE] FROM  DISK = N'\\oddo.fr\db_bck\TH3-DB-BCK\SQL\VOL_S2_MSB_SQLBCK_L21\XDATLWMUT0105$X_BOX_FR105_1\Database1\FULL\20210706\XDATLWMUT0105$X_BOX_FR105_1_Database1_FULL_20210706_222601_1.bak',  WITH  FILE = 1,  NORECOVERY,  NOUNLOAD,  STATS = 5
RESTORE LOG [TODELETE] FROM  DISK = N'\\oddo.fr\db_bck\TH3-DB-BCK\SQL\VOL_S2_MSB_SQLBCK_L21\XDATLWMUT0105$X_BOX_FR105_1\Database1\LOG\20210707\XDATLWMUT0105$X_BOX_FR105_1_Database1_LOG_20210707_004018.trn' WITH  FILE = 1,  NORECOVERY,  NOUNLOAD,  STATS = 5
RESTORE LOG [TODELETE] FROM  DISK = N'\\oddo.fr\db_bck\TH3-DB-BCK\SQL\VOL_S2_MSB_SQLBCK_L21\XDATLWMUT0105$X_BOX_FR105_1\Database1\LOG\20210707\XDATLWMUT0105$X_BOX_FR105_1_Database1_LOG_20210707_011700.trn' WITH  FILE = 1,  NOUNLOAD,  STATS = 5, STOPAT='07/07/2021 01:17:15'

Actual Behavior
RESTORE DATABASE [TODELETE] FROM  DISK = N'\\oddo.fr\db_bck\TH3-DB-BCK\SQL\VOL_S2_MSB_SQLBCK_L21\XDATLWMUT0105$X_BOX_FR105_1\Database1\FULL\20210706\XDATLWMUT0105$X_BOX_FR105_1_Database1_FULL_20210706_222601_1.bak',  WITH  FILE = 1,  NORECOVERY,  NOUNLOAD,  STATS = 5
RESTORE LOG [TODELETE] FROM  DISK = N'\\oddo.fr\db_bck\TH3-DB-BCK\SQL\VOL_S2_MSB_SQLBCK_L21\XDATLWMUT0105$X_BOX_FR105_1\Database1\LOG\20210707\XDATLWMUT0105$X_BOX_FR105_1_Database1_LOG_20210707_011700.trn' WITH  FILE = 1,  NOUNLOAD,  STATS = 5, STOPAT='07/07/2021 01:17:15'

**Comment by saxet77** [comment]
Expected Behavior generate from the SSMS 'Database/Restore/Script'

Actual behavior generated by DBATools (seems missing 1 log file)

**Comment by saxet77** [comment]
![Capture d’écran 2021-07-08 à 15 20 56](https://user-images.githubusercontent.com/29759165/124929060-7b688000-e000-11eb-875b-643aaa769d81.png)

Include the msdb..backupset table extract for the database/log backup LSN. 
DatabaseBackupLSN are the same for the backup database and the missing backup log.
Peharps it's the problem ...

**Comment by andreasjordan** [comment]
Currently I can not reproduce this. I just had a look at the code: BackupDatabaseLSN is not used.

**Comment by andreasjordan** [comment]
Does it return the correct script if you don't use -RestoreTime?

**Comment by andreasjordan** [comment]
If you look at $ListBackupFile - the file is missing already there - correct?

**Comment by andreasjordan** [comment]
Sorry for all the messages - I think I found the problem...

**Comment by andreasjordan** [comment]
@saxet77 - Can you please verify that you get the correct result from `Get-DbaDbBackupHistory` if you add `-IncludeCopyOnly`, so using `$ListBackupFile = Get-DbaDbBackupHistory -SqlInstance Server1 -Database Database1 -last -IncludeCopyOnly`?

@Stuart-Moore - Can you tell me more about this part?
https://github.com/sqlcollaborative/dbatools/blob/d81f07cc176102dda4c48ae620953d0f62ed3cb6/functions/Get-DbaDbBackupHistory.ps1#L340-L345
Especially the difference in the `Where-Object`: `[bigint]$_.DatabaseBackupLSN -eq [bigint]$fullDb.CheckPointLSN`.

**Comment by andreasjordan** [comment]
About your situation:
The full backup finished after the log backup started, but before the log backup finished. That's why the missing log backup has a different DatabaseBackupLSN than the other log backup and then CheckpointLSN of the full backup.
Because of that, the missing log backup is not been taken into account.

**Comment by saxet77** [comment]
@andreasjordan to answer your questions :
- Without the -RestoreTime same problem
- The missing log file is present in the $ListBackupFile
-  Same problem with -IncludeCopyOnly

**Comment by saxet77** [comment]
@andreasjordan, I will use -IncludeCopyOnly in my script.

You think it's need a fix on the Get-DbaDbBackupHistory cmdlet ?

**Comment by andreasjordan** [comment]
Yes, we have to fix Get-DbaDbBackupHistory. Your case (first log backup started before full finishes but finishes after full) is a rare case and thus has not seen in tests until now.
I don't understand the code difference between having IncludeCopyOnly set and not set yet. Buit I see using `-IncludeCopyOnly` as a workaround for you and I will try to find time to dig deep into that or maybe @Stuart-Moore has time as he is the author of that command.

**Comment by saxet77** [comment]
@andreasjordan 

I remade my tests and put you on the wrong issue:
- Get-DbaDbBackuphistory always return the "missing" log file with or without -IncludeCopyOnly parameter

but error is in the restore script generation command :

$RestoreScript = $ListBackupFile | Restore-DbaDatabase -SqlInstance YDATLWTST0030 -DatabaseName TODELETE -RestoreTime '2021-07-07 01:17:15' -WithReplace -OutputScriptOnly -ReplaceDbNameInFile -TrustDbBackupHistory;

Issue seems in Restore-DbaDatabase ...

Very sorry 
![Capture d’écran 2021-07-09 à 09 21 31](https://user-images.githubusercontent.com/29759165/125039755-20ce3300-e097-11eb-8fe3-b47ef0d7ea35.png)

**Comment by andreasjordan** [comment]
Can you add `-StopAfterGetBackupInformation -GetBackupInformation backupInformation` to the command? Then the variable `$backupInformation` is filled with the backups that would be restored.
I would like to know if the "missing log backup" is included there or not. So to have an idea where it gets filtered out.

**Comment by saxet77** [comment]
the $backupinformation variable own the "missing" log file take at 20210707_004018
<img width="1622" alt="Screenshot 2021-07-09 at 14 21 42" src="https://user-images.githubusercontent.com/29759165/125077496-74ee0d00-e0c1-11eb-9345-bb45ccbe211f.png">

**Comment by andreasjordan** [comment]
So here the $FilteredBackupHistory (that is what is in your $backupinformation) is still including the "missing log backup":
https://github.com/sqlcollaborative/dbatools/blob/879021234ae8730f4550d75f1ee43397b6b673a7/functions/Restore-DbaDatabase.ps1#L705

`Test-DbaBackupInformation` is then building an internal `$DbHistory` which is passed to `Test-DbaLsnChain` and failes there because of the missing log backup.

Ok, I'll have a deeper look into `Test-DbaBackupInformation`...

**Comment by andreasjordan** [comment]
Can you please run the command that fails with `-Verbose` - maybe that output helps us.

**Comment by saxet77** [comment]
$RestoreScript = $ListBackupFile | Restore-DbaDatabase -SqlInstance YDATLWTST0030 -DatabaseName TODELETE -RestoreTime '2021-07-07 01:17:15' -WithReplace -OutputScriptOnly -ReplaceDbNameInFile -TrustDbBackupHistory -Verbose; #-StopAfterGetBackupInformation -GetBackupInformation backupInformation 


[restore-dbadatabase.verbose.txt](https://github.com/sqlcollaborative/dbatools/files/6792439/restore-dbadatabase.verbose.txt)

**Comment by andreasjordan** [comment]
Thanks for the verbose output - but I still don't see where the log backup gets filtered out. We have to wait for @Stuart-Moore - maybe he has an idea.

**Comment by Stuart-Moore** [comment]
@andreasjordan  
DatabaseBackupLSN is the LSN of the last 'real' full backup taken before the current backup was taken. 

CheckpointLSN is usuallly the same as FirstLSN and is the LSN to which the backup would be restored to. It's rare to not have them be the same but it can happen

Reading those 2 lines I can't remember why they're different. Again, i expect it was another edge case that cropped up

@saxet77 is it possible for you to share you backup history? Then we can wrap it into the tests to make it easier to play with/replicate

If you could just run:

<your get history command> | Export-CliXml -Path c:\somewhere\file.xml

and upload it to here, then we can rehydrate it and work with. If you want to obfuscate server and db names that's fine. Really just want the LSNs and the other metadata.

**Comment by Stuart-Moore** [comment]
sorry, mucked up the Markdown in the last comment. Pleae could you upload:
`$ListBackupFile |  Export-CliXml -Depth 5 -Path c:\somewhere\file.xml`

I can fake something if needed, but having 'real' data will make sure we have a proper test to stop it coming back again

**Comment by saxet77** [comment]
Hi Stuart,
No pbl, Let me know if you have all information to your test in the join file.

[ListBackupFile.zip](https://github.com/sqlcollaborative/dbatools/files/6797260/ListBackupFile.zip)

**Comment by Stuart-Moore** [comment]
Going go need a bit more info please, was concentrating on the restore side to much. Could you dump out the raw backup history for me please? Will need something like:
`Get-DbaDbBackupHistory -sqlinstance x -database y -since '10/07/2021 21:10' | Export-CliXml -path -depth 5 ./some.xml`

Sorry for the delay, time is being eaten up by a large migration atm.

**Comment by saxet77** [comment]
Hi,

Find the compressed .xml file asked from the following command:
Get-DbaDbBackupHistory -sqlinstance XDATLWMUT0105 -database DB_LDW_BHF -since '2021-07-10 21:10:00' | Export-CliXml -depth 5 -path D:\Temp\ListBackupFile.20210714.xml

[ListBackupFile.20210714.zip](https://github.com/sqlcollaborative/dbatools/files/6818296/ListBackupFile.20210714.zip)

**Comment by Stuart-Moore** [comment]
Think i now have a fix for this, but it's failing a single test.  Need to rebuild my local test rig to work out why.

**Comment by Stuart-Moore** [comment]
Ok, so this is getting annoying. Passes fine locally:
![image](https://user-images.githubusercontent.com/5613670/126518581-1a1cddc8-996b-4bdb-b4df-4db890076e42.png)

But fails in Appveyor. 

Trying to rdp in and it appears to not be liking the password.  Is it not just the one in appveyor.yml anymore? (@andreasjordan  or @potatoqualitee  can you help here?)

**Comment by potatoqualitee** [comment]
oh sorry, ill msg you on twitter, stuart

**Comment by Stuart-Moore** [comment]
Gah, it's a bug with 2008r2! That's why it works fine for me.

For some reason 2008r2 returns 3 rows of history, others return 4. Both result sets are restoreable though, so that's all that really matters.  will fix up tests and push it back up

**Comment by potatoqualitee** [comment]
haha awesome work! 🚒

**Comment by Stuart-Moore** [comment]
Yeah, that was confusing the heck out of me. The good news is that the filtering of all back up history is done by select-dbabackupinformation, so we should have consistency between get-dbadbbackuphistory and the restore stack.

@saxet77 Really sorry for the delay on this one.  Would you be able to take this new branch for a spin and see if it fixes your issue. I think it should, but would be great to know it does.

**Comment by potatoqualitee** [comment]
Thanks, @Stuart-Moore ! I'll wait for your PR to release 1.1.0 💯

**Comment by saxet77** [comment]
Hi @Stuart-Moore, no problem for the delay.
Not very familiar with githup "how can I take this new branch" to validate the fix on my side ?
Thks

**Comment by Stuart-Moore** [comment]
I can walk you  through it. Do you have any git experience? You won't need much but just means i know where to target things

**Comment by potatoqualitee** [comment]
Is the fix in the dev branch? if so, it can be downloaded from https://dbatools.io/devzip

**Comment by Stuart-Moore** [comment]
Not yet. Worst case I can share a one drive link to a zip of the lsnbreak branch.

**Comment by potatoqualitee** [comment]
Ahh there's the branch! Here's the zip

https://github.com/sqlcollaborative/dbatools/archive/refs/heads/lsnbreak.zip

**Comment by saxet77** [comment]
Thks, got the zip file.
tests in progress ... ;)

**Comment by potatoqualitee** [comment]
sweet! looking forward to hearing back

**Comment by Stuart-Moore** [comment]
Great, and I’ve learnt a new Github feature as well 👍

Cheers
Stuart

**Comment by potatoqualitee** [comment]
Through the website, you can select the branch, then click on the green button and copy the link from Download zip 👍🏼

**Comment by saxet77** [comment]
running with the 1.0.173 version and the same problem ;(

PS C:\Users\fblondin_bst> get-module -name dbatools

ModuleType Version    Name                                ExportedCommands                                                                                                                           
---------- -------    ----                                ----------------                                                                                                                           
Script     1.0.173    dbatools                            {Select-DbaObject, Set-DbatoolsConfig, Add-DbaAgDatabase, Add-DbaAgListener...}                                                            



PS C:\Users\fblondin_bst>     $RestoreScript = $ListBackupFile | Restore-DbaDatabase -SqlInstance YDATLWTST0030 -DatabaseName TODELETE -RestoreTime '2021-07-07 01:17:15' -WithReplace -OutputScriptOnly -ReplaceDbNameInFile -TrustDbBackupHistory #-Verbose; #-StopAfterGetBackupInformation -GetBackupInformation backupInformation 


 PS C:\Users\fblondin_bst> get-module -name dbatools

ModuleType Version    Name                                ExportedCommands                                                                                                                           
---------- -------    ----                                ----------------                                                                                                                           
Script     1.0.173    dbatools                            {Select-DbaObject, Set-DbatoolsConfig, Add-DbaAgDatabase, Add-DbaAgListener...}                                                            



PS C:\Users\fblondin_bst>     $RestoreScript = $ListBackupFile | Restore-DbaDatabase -SqlInstance YDATLWTST0030 -DatabaseName TODELETE -RestoreTime '2021-07-07 01:17:15' -WithReplace -OutputScriptOnly -ReplaceDbNameInFile -TrustDbBackupHistory #-Verbose; #-StopAfterGetBackupInformation -GetBackupInformation backupInformation

**Comment by potatoqualitee** [comment]
hey @saxet77  - please close out powershell entirely then `Import-module C:\wherever\to\the\dir\dbatools.psd1 -Force`

**Comment by saxet77** [comment]
Hi @potatoqualitee ,

I followed your advice and remake my test, but same break LSN error
(providing an export of $ListBackupFile
[ListBackupFile.20210724.xml.txt](https://github.com/sqlcollaborative/dbatools/files/6872133/ListBackupFile.20210724.xml.txt)
)
# Get backup history from the SourceInstance msdb database
$ListBackupFile = Get-DbaDbBackupHistory -SqlInstance xxxxxxxx -Database xxxxxxx -Since '2021-07-06 21:00:00' | where end -lt '2021-07-07 02:30:00' 

# Generate the PIT restore script
$RestoreScript = $ListBackupFile | Restore-DbaDatabase -SqlInstance xxxxxxxx -DatabaseName TODELETE -RestoreTime '2021-07-07 01:17:15' -WithReplace -OutputScriptOnly -TrustDbBackupHistory ;

Error message:
[Error.20210720.txt](https://github.com/sqlcollaborative/dbatools/files/6872135/Error.20210720.txt)

**Comment by potatoqualitee** [comment]
bummer. thank you for the files, tho. they may be able to help stuart 💯

**Comment by Stuart-Moore** [comment]
Take 2. Would you mind taking this new version for a spin. It seems to work on the 'msdb' I've hacked together from the files, but need to know it works with a proper one please

https://github.com/sqlcollaborative/dbatools/archive/refs/heads/lsnbreak.zip

**Comment by saxet77** [comment]
Sorry ;(
Same error message with this new version.

**Comment by andreasjordan** [comment]
@Stuart-Moore Do you have an eye on this?

**Comment by andreasjordan** [comment]
Do we still want to try to solve this? Or is the root cause a very rare situation and not worth the effort? Will close this in a few days if there are no dissenting votes.

**Comment by potatoqualitee** [comment]
I'll defer to @Stuart-Moore

**Comment by potatoqualitee** [comment]
sometimes stuart gets free cycles, let's hold off to closed this until that time comes up.

**Comment by andreasjordan** [comment]
I will close thi now, but can reopen if needed.

**Comment by wsmelton** [comment]
This is still a known issue and should not be closed.

**Comment by andreasjordan** [comment]
I finally can reprodure the issue (update: only once and havn't saved it) and will start working on this.

Here is my script (that does not work):
```
$sqlInstance = 'SQL01'
$database = 'BackupTest7515'

$null = New-DbaDatabase -SqlInstance $sqlInstance -Name $database 
Invoke-DbaQuery -SqlInstance $sqlInstance -Database $database -Query "CREATE TABLE JustWasteSpace (id int IDENTITY PRIMARY KEY, space CHAR(100) DEFAULT 'SPACE')"
Invoke-DbaQuery -SqlInstance $sqlInstance -Database $database -Query "INSERT INTO JustWasteSpace DEFAULT VALUES"

$null = New-DbaAgentJob -SqlInstance $sqlInstance -Job "$database-Full"
$null = New-DbaAgentJobStep -SqlInstance $sqlInstance -Job "$database-Full" -StepName Backup -Subsystem TransactSql -Command "BACKUP DATABASE $database TO DISK='$database.bak'; INSERT INTO $database.dbo.JustWasteSpace DEFAULT VALUES; CHECKPOINT"
$null = New-DbaAgentJob -SqlInstance $sqlInstance -Job "$database-Log"
$null = New-DbaAgentJobStep -SqlInstance $sqlInstance -Job "$database-Log" -StepName Backup -Subsystem TransactSql -Command "CHECKPOINT; BACKUP LOG $database TO DISK='$database.trn'"
#$null = New-DbaAgentJob -SqlInstance $sqlInstance -Job "$database-Insert"
#$null = New-DbaAgentJobStep -SqlInstance $sqlInstance -Job "$database-Insert" -StepName Insert -Subsystem TransactSql -Database $database -Command "INSERT INTO JustWasteSpace DEFAULT VALUES; CHECKPOINT; WAITFOR DELAY '00:00:01';INSERT INTO JustWasteSpace DEFAULT VALUES; CHECKPOINT; WAITFOR DELAY '00:00:01';INSERT INTO JustWasteSpace DEFAULT VALUES; CHECKPOINT; WAITFOR DELAY '00:00:01';INSERT INTO JustWasteSpace DEFAULT VALUES; CHECKPOINT; WAITFOR DELAY '00:00:01';INSERT INTO JustWasteSpace DEFAULT VALUES; CHECKPOINT; WAITFOR DELAY '00:00:01';INSERT INTO JustWasteSpace DEFAULT VALUES; CHECKPOINT; WAITFOR DELAY '00:00:01';INSERT INTO JustWasteSpace DEFAULT VALUES; CHECKPOINT;"
#$null = New-DbaAgentSchedule -SqlInstance $sqlInstance -Job "$database-Insert" -Schedule "$database-Insert" -FrequencyType Daily -FrequencyInterval 1 -FrequencySubdayType Second -FrequencySubdayInterval 10 -Force

$null = Start-DbaAgentJob -SqlInstance $sqlInstance -Job "$database-Full" -Wait
$null = Start-DbaAgentJob -SqlInstance $sqlInstance -Job "$database-Log" -Wait

1..26 | % { Invoke-DbaQuery -SqlInstance $sqlInstance -Database $database -Query "INSERT INTO JustWasteSpace SELECT space FROM JustWasteSpace" }

# Now we have a log of data to back up, both in the data file and the log file
# Goal is that the log backup starts while full backup is running and finishes after the full backup finished
$null = Start-DbaAgentJob -SqlInstance $sqlInstance -Job "$database-Full"
Start-Sleep -Seconds 3
Invoke-DbaQuery -SqlInstance $sqlInstance -Database $database -Query "INSERT INTO JustWasteSpace DEFAULT VALUES"
$null = Start-DbaAgentJob -SqlInstance $sqlInstance -Job "$database-Log"

# Wait for the backups to finish and start one more log backup
Start-Sleep -Seconds 30
Invoke-DbaQuery -SqlInstance $sqlInstance -Database $database -Query "INSERT INTO JustWasteSpace DEFAULT VALUES"
$null = Start-DbaAgentJob -SqlInstance $sqlInstance -Job "$database-Log" -Wait


# get the backup history (complete and last)
$fullHistory = Get-DbaDbBackupHistory -SqlInstance $sqlInstance -Database $database 
$fullHistory | ogv

$lastHistory = Get-DbaDbBackupHistory -SqlInstance $sqlInstance -Database $database -Last
$lastHistory | ogv

# try to use the last history to restore the database
$restore = $lastHistory | Restore-DbaDatabase -SqlInstance $sqlInstance -DatabaseName "$database-Restore" -ReplaceDbNameInFile



# cleanup


$null = Remove-DbaAgentJob -SqlInstance $sqlInstance -Job "$database-Full" -Confirm:$false
$null = Remove-DbaAgentJob -SqlInstance $sqlInstance -Job "$database-Log" -Confirm:$false
#$null = Remove-DbaAgentJob -SqlInstance $sqlInstance -Job "$database-Insert" -Confirm:$false
$null = Remove-DbaDbBackupRestoreHistory -SqlInstance $sqlInstance -Database $database -Confirm:$false
$null = Remove-DbaDatabase -SqlInstance $sqlInstance -Database $database -Confirm:$false
$null = Remove-DbaDbBackupRestoreHistory -SqlInstance $sqlInstance -Database "$database-Restore" -Confirm:$false
$null = Remove-DbaDatabase -SqlInstance $sqlInstance -Database "$database-Restore" -Confirm:$false
```


Goal is to have a situation where the end of the log backup is after the end of the full backup. This works, but it also needs to have a higher end-lsn than the full backup. And this does not work.

Have to stop for today, but will work again on this in the next day.

If anyone can help in setting up this situation, I would be glad to hear from you.

**Comment by andreasjordan** [comment]
> ![Capture d’écran 2021-07-08 à 15 20 56](https://user-images.githubusercontent.com/29759165/124929060-7b688000-e000-11eb-875b-643aaa769d81.png)
> 
> Include the msdb..backupset table extract for the database/log backup LSN. DatabaseBackupLSN are the same for the backup database and the missing backup log. Peharps it's the problem ...

After spending some more time on this I think this picture is a key to understanding the problem.

Let me describe the situation and kind of shorten the LSN numbers for better readability:
* The full backup starts at LSN 449 and ends at LSN 464, and has a checkpoint LSN 449 (!)
* The first log backup starts at LSN 461 and ends at LSN 467 - so it overlaps the end of the full backup both in time and LSN. (I still can not reproduce this situation). As the database backup LSN it has the checkpoint LSN of a privious full backup. I agree with the author of this issue that this is the key, will go into detail later.
* The second log backup starts at LSN 467 and ends at LSN 485. The database backup LSN is 449 and is equal to the checkpoint LSN of the full backup.

As @Stuart-Moore and also https://sqlbak.com/academy/backup-chain describe it, the database backup LSN tell us, to which full backup the log backup belongs. So this is coded in line 361 of Get-DbaDbBackupHistory: `[bigint]$_.DatabaseBackupLSN -eq [bigint]$fullDb.CheckPointLSN`. This is the reason that the first log backup is not beeing restored, although the last LSN is greater than the last LSN of the full backup and thus has transactions that need to be applied.

I managed to manipulate the database backup LSN of the first log backup in my lab and this way the first log backup is not returnd by the command. But I also tested with SSMS and SSMS still returns the first log backup in the list of needed restores. But SSMS only gets all the entries from msdb..backupset and so I don't see the code that returns the list. Maybe they only test based on first LSN and last LSN.

If we want to do it like SSMS, then we would need to speak with a nice person at Microsoft and ask for help.

Or we say that the database backup LSN of the first log backup is wrong (= it's a bug in SQL Server - that we can not reproduce) and our code is right.

Maybe this helps the other contributors (@Stuart-Moore / @wsmelton / @potatoqualitee ) to decide what to do.

**Comment by andreasjordan** [comment]
One more update: I could reproduce half of the situation: The first log backup has a lower database backup LSN than expected, but the same end LSN as the full backup. So this log backup is not needed for the restore.

As expected dbatools only uses the second log backup and can restore that without problem. SSMS uses first log backup as well, which is not needed and only extra time spend - but without any error, so use would not notice.

Here are the screenshots:
![image](https://github.com/dataplat/dbatools/assets/66946165/c212955b-5e29-4cf5-bc00-6c4407897c44)

![image](https://github.com/dataplat/dbatools/assets/66946165/6ee5d5e2-e1db-45a5-acd2-bd4cc1a5b068)

![image](https://github.com/dataplat/dbatools/assets/66946165/339c1469-cd93-4489-a56d-8dca1f04f936)

So do we want the same filtering as SSMS? Maybe they filter on last LSN of log backup must be greater or equal than last LSN of full backup.

**Comment by wsmelton** [comment]
You can trace the t-sql that SSMS is executing if you can get the same backup result set.  Microsoft is not hiding it from view.

**Comment by andreasjordan** [comment]
I did start a trace, but found nothing useful. So they do it in the application. Which makes sense as I can select and de-select individual backups. So they only get the complete history once and then slice and dice in the GUI.

**Comment by katamoto75** [comment]
We have the same issue with the missing last log file in `Get-DbaDbBackupHistory`. This is because we offloaded backups to our secondary replica and had to use `copy_only` for full backups. In this case, `DatabaseBackupLSN` never updates and is always less than `CheckpointLSN`. I discovered this while using `Restoring-DbaDatabase` with `-RestoreTime` and not getting all the expected data.

**Comment by AdamJKoehler** [comment]
We have this same problem where a log backup that is needed for restore is tagged as a break in the LSN Chain.  `Restore-DbaDatabase `fails this process, but if we go through the process described in` Invoke-DbaAdvancedRestore` and use the pipeline for the restore, it works without a problem.  I found the problem to be down in `Test-DBALsnChain` on line 99.  It needs additional checks to see if the FirstLSN of the log backup being checked is greater than the FULL backup's CheckPointLSN.

**Comment by AdamJKoehler** [comment]
> We have this same problem where a log backup that is needed for restore is tagged as a break in the LSN Chain. `Restore-DbaDatabase`fails this process, but if we go through the process described in` Invoke-DbaAdvancedRestore`and use the pipeline for the restore, it works without a problem. I found the problem to be down in `Test-DBALsnChain`on line 99. It needs additional checks to see if the FirstLSN of the log backup being checked is greater than the FULL backup's CheckPointLSN.
> 
> Our scenario is just like what's referenced in Brent Ozar's link above.

@wsmelton @potatoqualitee @Stuart-Moore fyi...

**Comment by andreasjordan** [comment]
Will (hopefully) be solved with #10201 / #10185.
